language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/FixReturnTypeCodeAction.java

```diff
@@ -100,20 +106,62 @@ public List<CodeAction> getDiagBasedCodeActions(Diagnostic diagnostic,
         }
 
         List<TextEdit> importEdits = new ArrayList<>();
-        List<String> types = new ArrayList<>();
+        List<Set<String>> types = new ArrayList<>();
+        List<CodeAction> codeActions = new ArrayList<>();
         boolean returnTypeDescPresent = funcDef.get().functionSignature().returnTypeDesc().isPresent();
 
         if (checkExprDiagnostic) {
             // Add error return type for check expression
             if (returnTypeDescPresent) {
-                types.add(funcDef.get().functionSignature().returnTypeDesc().get().type().toString().trim().concat("|")
-                        .concat("error"));
+                types.add(Collections.singleton(funcDef.get().functionSignature().returnTypeDesc().get().type()
+                        .toString().trim().concat("|").concat("error")));
             } else {
-                types.add("error?");
+                types.add(Collections.singleton("error?"));
             }
         } else {
-            // Get all possible return types including ambiguous scenarios
-            types = CodeActionUtil.getPossibleTypes(foundType.get(), importEdits, context);
+            List<List<String>> combinations = new ArrayList<>();
+            ReturnStatementFinder returnStatementFinder = new ReturnStatementFinder();
+            returnStatementFinder.visit(funcDef.get());
+            List<ReturnStatementNode> nodeList = returnStatementFinder.getNodeList();
+
+            for (ReturnStatementNode returnStatementNode : nodeList) {
+                if (returnStatementNode.expression().isEmpty() || context.currentSemanticModel().isEmpty()) {
+                    return Collections.emptyList();
+                }
+                ExpressionNode expression = returnStatementNode.expression().get();
+                SemanticModel semanticModel = context.currentSemanticModel().get();
+                Optional<TypeSymbol> typeSymbol = semanticModel.typeOf(expression);
+                if (typeSymbol.isEmpty() || typeSymbol.get().typeKind() == TypeDescKind.COMPILATION_ERROR) {
+                    return Collections.emptyList();
+                }
+                if (typeSymbol.get().typeKind() == TypeDescKind.FUNCTION) {
+                    combinations.add(Collections.singletonList("(" + CodeActionUtil.getPossibleTypes(typeSymbol.get(),
+                            importEdits, context).get(0) + ")"));
+                } else {
+                    combinations.add(CodeActionUtil.getPossibleTypes(typeSymbol.get(), importEdits, context));
+                }
+            }
+
+            for (List<String> stringTypes : combinations) {
```

Comment:
Shall we extract this logic to a separate method and add comments explaining key steps

Reply:
Done